### DISCUSSION TITLE
Custom callback functions for IPOPT NLP objective and constraints not working in C++

### DISCUSSION THREAD

**Discussion by dhathris** [casadi/casadi#3791] [Q&A]
`class SKObjectiveCallback : public Callback {
        public:
            SKObjectiveCallback(const std::string& name, const Dict& opts) {
                construct(name, opts);
            }
            ~SKObjectiveCallback() override {}

            casadi_int get_n_in() override {
                return input_.size();
            }

            casadi_int get_n_out() override {
                return 1;
            }

            void init() override {

            }

            std::vector<DM> eval(const std::vector<DM>& arg) const override;            

        private:
            VectorXd input_;
            int N_;
    };`

Above is the way I declared my own callback for the objective function for IPOPT NLP solver. I have a similar definition for constraints.
When I instantiate these custom callback, the program runs into a segmentation fault.

`SKObjectiveCallback obj("Objective", opts);`

What could be the possible root causes of this behavior?

**Comment by jgillis** [comment]
Make sure you keep obj in scope. casadi does *not* take ownership.

**Reply by dhathris** [reply]
The problem is that the instance `obj` is not getting created, the call to this constructor itself is throwing a segmentation fault.

**Reply by dhathris** [reply]
For future reference for those who might encounter this same problem. get_n_in and get_n_out methods expected a return of a constant value. In this example, the value returned by input_.size() in get_n_in method is a variable, that's why I got a segmentation fault.
